### PR TITLE
Add ability to configure azure provisioner credential through AzureProvisionerOptions

### LINKED ISSUE #1329: [Azure Provisioner] Allow configuration for DefaultAzureCredentialOptions used by AddAzureProvisioning
It seems that currently the `AddAzureProvisioning` is creating a new defaultAzureCredential with some predefined options [here](https://github.com/dotnet/aspire/blob/2b6951810bc781e5ff71edd5815b4fd2cba05a7e/src/Aspire.Hosting.Azure.Provisioning/Provisioners/AzureProvisioner.cs#L69)

However, that is blocking me from using a different subscription (e.g. my personal one where I have VS Enterprise credits) as the default one will be loaded from VS signed-in account.

### PR DESCRIPTION
**Pull request by oising**
Fixes #1329

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/3083)

### CODE REVIEW

File: src/Aspire.Hosting.Azure/Provisioning/AzureProvisionerOptions.cs [hunk 1]
```diff
@@ -12,4 +14,10 @@ internal sealed class AzureProvisionerOptions
     public bool? AllowResourceGroupCreation { get; set; }
 
     public string? Location { get; set; }
+
+    [AllowedValues([
+        "AzureCli", "AzurePowerShell", "VisualStudio", "VisualStudioCode",
+        "AzureDeveloperCli", "InteractiveBrowser", "Default"
+    ])]
+    public string? ProvisionerCredentialSource { get; set; } = "Default";
```
Comment (davidfowl) [inline src/Aspire.Hosting.Azure/Provisioning/AzureProvisionerOptions.cs:22]:
```suggestion
    public string CredentialSource { get; set; } = "Default";
```

**Review by davidfowl — APPROVED** [review]

File: src/Aspire.Hosting.Azure/Provisioning/AzureProvisionerExtensions.cs [hunk 2]
```diff
@@ -26,7 +26,9 @@ public static IDistributedApplicationBuilder AddAzureProvisioning(this IDistribu
 
         // Attempt to read azure configuration from configuration
         builder.Services.AddOptions<AzureProvisionerOptions>()
-            .BindConfiguration("Azure");
+            .BindConfiguration("Azure")
```
Comment (mitchdenny) [inline src/Aspire.Hosting.Azure/Provisioning/AzureProvisionerExtensions.cs:29]: OK with this for now but might we want this to be configurable @davidfowl?

File: src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureProvisioner.cs [hunk 3]
```diff
@@ -275,13 +275,36 @@ private async Task ProcessResourceAsync(IConfiguration configuration, Lazy<Task<
 
     private async Task<ProvisioningContext> GetProvisioningContextAsync(Lazy<Task<JsonObject>> userSecretsLazy, CancellationToken cancellationToken)
     {
-        var credential = new DefaultAzureCredential(new DefaultAzureCredentialOptions()
+        // Optionally configured in AppHost appSettings under "Azure" : { "CredentialSource": "AzureCli" }
+        var credentialSetting = _options.CredentialSource;
+
+        TokenCredential credential = credentialSetting switch
```
Comment (eerhardt) [inline src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureProvisioner.cs:281]: FYI @tg-msft. We could really use the generalized version of this from the Azure SDK .... 😉

File: src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureProvisioner.cs [hunk 4]
```diff
@@ -275,13 +275,36 @@ private async Task ProcessResourceAsync(IConfiguration configuration, Lazy<Task<
...
+        {
+            "AzureCli" => new AzureCliCredential(),
+            "AzurePowerShell" => new AzurePowerShellCredential(),
+            "VisualStudio" => new VisualStudioCredential(),
+            "VisualStudioCode" => new VisualStudioCodeCredential(),
+            "AzureDeveloperCli" => new AzureDeveloperCliCredential(),
+            "InteractiveBrowser" => new InteractiveBrowserCredential(),
+            _ => new DefaultAzureCredential(new DefaultAzureCredentialOptions()
```
Comment (mitchdenny) [inline src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureProvisioner.cs:289]: Wouldn't we want the fallback to try everything?

Comment (davidfowl) [reply src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureProvisioner.cs:289]: This was what we had before.

File: src/Aspire.Hosting.Azure/Provisioning/AzureProvisionerExtensions.cs (hunk 2 above)
Comment (davidfowl) [reply src/Aspire.Hosting.Azure/Provisioning/AzureProvisionerExtensions.cs:29]: Not in this PR. The provisioner already hardcodes this section.

**Pull request merged**